phpunit/class-rest-scripts-controller-test.php

```diff
@@ -0,0 +1,384 @@
+	}
+
+	/**
+	 * Test single script.
+	 */
+	public function test_get_item() {
+		foreach ( array( self::$admin_id, self::$superadmin_id ) as $user_id ) {
+			wp_set_current_user( $user_id );
+			$request  = new WP_REST_Request( 'GET', '/__experimental/scripts/' . self::$script_handle );
+			$response = rest_get_server()->dispatch( $request );
+			$data     = $response->get_data();
+
+			$this->assertEquals( self::$script_handle, $data['handle'] );
+			$this->assertEquals( home_url( '/test.js' ), $data['src'] );
+			$this->assertEquals( home_url( '/test.js' ), $data['url'] );
+		}
+	}
+
+	/**
+	 * Create item test.
+	 */
+	public function test_create_item() {
+		// Not testable.
+	}
+
+	/**
+	 * Update item test.
+	 */
+	public function test_update_item() {
+		// Not testable.
+	}
+
+	/**
+	 * Delete item test.
+	 */
+	public function test_delete_item() {
+		// Not testable.
+	}
+
+	/**
+	 * Test prepare item.
+	 */
+	public function test_prepare_item() {
+		wp_set_current_user( self::$admin_id );
+		$request  = new WP_REST_Request( 'GET', '/__experimental/scripts/' . self::$script_handle );
+		$response = rest_get_server()->dispatch( $request );
+		$data     = $response->get_data();
+
+		$this->assertArrayHasKey( 'handle', $data );
+		$this->assertArrayHasKey( 'src', $data );
+		$this->assertArrayHasKey( 'url', $data );
+		$this->assertArrayHasKey( 'args', $data );
+		$this->assertArrayHasKey( 'ver', $data );
+		$this->assertArrayHasKey( 'extra', $data );
+		$this->assertArrayHasKey( 'textdomain', $data );
+		$this->assertArrayHasKey( 'translations_path', $data );
+		$this->assertArrayHasKey( 'deps', $data );
+		$links = $response->get_links();
+		$this->assertArrayHasKey( 'self', $links );
+		$this->assertArrayHasKey( 'collection', $links );
+		$this->assertArrayHasKey( 'deps', $links );
+	}
+
+	/**
+	 * Test schema.
+	 */
+	public function test_get_item_schema() {
+		$request  = new WP_REST_Request( 'OPTIONS', '/__experimental/scripts' );
+		$response = rest_get_server()->dispatch( $request );
+		$data     = $response->get_data();
+		$this->assertArrayHasKey( 'schema', $data );
+		$this->assertArrayHasKey( 'properties', $data['schema'] );
+		$properties = $data['schema']['properties'];
+		$this->assertEquals( 9, count( $properties ) );
+		$this->assertArrayHasKey( 'handle', $properties );
+		$this->assertArrayHasKey( 'src', $properties );
+		$this->assertArrayHasKey( 'url', $properties );
+		$this->assertArrayHasKey( 'args', $properties );
+		$this->assertArrayHasKey( 'ver', $properties );
+		$this->assertArrayHasKey( 'extra', $properties );
+		$this->assertArrayHasKey( 'textdomain', $properties );
+		$this->assertArrayHasKey( 'translations_path', $properties );
+		$this->assertArrayHasKey( 'deps', $properties );
+	}
+
+	/**
+	 * Test single block script.
+	 */
+	public function test_get_item_block_script() {
+		foreach ( array( 0, self::$subscriber_id, self::$editor_id, self::$author_id, self::$admin_id, self::$superadmin_id ) as $user_id ) {
```

Comment:
I don't personally like this style of loop. Can we just use a dataprovider here and other places that this format is used? Using dataprovider really helps in debugging.

Reply:
@spacedmonkey I tried to use data providers and they were not delivering the right data - if they return primitives like integers then they work fine. When I tried to return class props they were always set to null. Any idea what I did incorrectly?